langkit/templates/pkg_rewriting_impl_body_ada.mako

```diff
@@ -27,11 +27,132 @@ package body ${ada_lib_name}.Rewriting_Implementation is
+      )
+      %>
+      ## Check that `index` is in range 1 .. `upper_bound`, where
+      ## `upper_bound` is equal to the number of children of `handle`
+      ## if `for_insertion` is false, or to that number + 1 if it is true.
+      ##
+      ## Note: the call to Pre_Check is inlined so that the construction
+      ## of the exception message happens only if the check fails.
+      if ${index} > ${upper_bound} then
+         raise Precondition_Failure
+            with "Invalid index " & ${index}'Image & ": ${handle} has " &
+                  Children_Count (${handle})'Image & " children";
+      end if;
+   </%def>
+
+   <%def name="pre_check_unit_no_diags(unit)">
+      ## Check that the given analysis unit does not have any diagnostics
+      Pre_Check
+        (not Has_Diagnostics (${unit}),
+         "${unit} must not have diagnostics");
+   </%def>
+
+   <%def name="pre_check_is_token_kind(kind)">
+      ## Check that the given node kind corresponds to a token node kind
+      Pre_Check
+        (Is_Token_Node (${kind}),
+         "Expected a token node. Got " & ${kind}'Image);
+   </%def>
+
+   <%def name="pre_check_is_not_token_kind(kind)">
+      ## Check that the given node kind *does not* correspond to a token node
+      ## kind.
+      Pre_Check
+        (not Is_Token_Node (${kind}),
+         "Expected a token node. Got " & ${kind}'Image);
+   </%def>
+
+   <%def name="pre_check_is_list_kind(kind)">
+      ## Check that the given node kind corresponds to a list node kind
+      Pre_Check
+        (Is_List_Node (${kind}),
+         "Expected a list node. Got " & ${kind}'Image);
+   </%def>
+
+   <%def name="pre_check_null_or_untied(handle)">
+      ## Check that the given node handle is not tied to any unit rewriting
+      ## handle.
+      ## Note: No_Node_Rewriting_Handle matches this predicate.
+      Pre_Check
+        (${handle} = No_Node_Rewriting_Handle or else not Tied (${handle}),
+         "${handle} must not be tied to another rewriting context.");
+   </%def>
+
+   <%def name="pre_check_is_tied(handle)">
+      ## Check that the given node rewriting handle is well tied to an analysis
+      ## unit.
+      Pre_Check
+        (Tied (${handle}),
+         "${handle} must be tied to an analysis unit.");
+   </%def>
+
+   <%def name="pre_check_null_or_valid_context(node, handle)">
+      ## Check that the given node rewriting handle `node` is either null, or is
+      ## associated to the given rewriting handle `handle`.
+      Pre_Check
+        (${node} = No_Node_Rewriting_Handle
+            or else Context (${node}) = ${handle},
+         "${node} should be associated to rewriting context ${handle}.");
+   </%def>
+
    function Handle (Context : Internal_Context) return Rewriting_Handle is
      (Convert (Get_Rewriting_Handle (Context)));
 
    function Context (Handle : Rewriting_Handle) return Internal_Context is
-     (Handle.Context);
+   begin
+      ${pre_check_rw_handle("Handle")}
```

Comment:
Except in `<%def>` directives, please replace all double quotes with single ones (I try to keep this use consistent unless for string literals that must contain single quotes).